GVFS/GVFS/CommandLine/DehydrateVerb.cs

```diff
@@ -145,31 +164,139 @@ protected override void Execute(GVFSEnlistment enlistment)
                 // Local cache and objects paths are required for TryDownloadGitObjects
                 this.InitializeLocalCacheAndObjectsPaths(tracer, enlistment, retryConfig, serverGVFSConfig: null, cacheServer: null);
 
-                if (this.TryBackupFiles(tracer, enlistment, backupRoot))
+                if (string.IsNullOrEmpty(this.Folders))
                 {
-                    if (this.TryDownloadGitObjects(tracer, enlistment, retryConfig) &&
-                        this.TryRecreateIndex(tracer, enlistment))
+                    this.RunFullDehydrate(tracer, enlistment, backupRoot, retryConfig);
+                }
+                else
+                {
+                    string[] folders = this.Folders.Split(new[] { ";" }, StringSplitOptions.RemoveEmptyEntries);
+
+                    if (folders.Length > 0)
                     {
-                        // Converting the src folder to partial must be the final step before mount
-                        this.PrepareSrcFolder(tracer, enlistment);
-                        this.Mount(tracer);
+                        if (cleanStatus)
+                        {
+                            this.DehydrateFolders(tracer, enlistment, backupRoot, folders);
+                        }
+                        else
+                        {
+                            this.ReportErrorAndExit("Must have a clean git status to dehydrate folders.");
+                        }
+                    }
+                    else
+                    {
+                        this.ReportErrorAndExit("No folders to dehydrate.");
+                    }
+                }
+            }
+        }
+
+        private void DehydrateFolders(JsonTracer tracer, GVFSEnlistment enlistment, string backupRoot, string[] folders)
+        {
+            List<string> foldersToDehydrate = new List<string>();
+
+            if (!this.ShowStatusWhileRunning(
+                () =>
+                {
+                    string ioError;
+                    foreach (string folder in folders)
+                    {
+                        string fullPath = Path.Combine(enlistment.WorkingDirectoryRoot, folder);
+                        string backupPath = Path.Combine(backupRoot, folder);
+                        if (this.fileSystem.DirectoryExists(fullPath))
+                        {
+                            if (!this.TryIO(tracer, () => this.fileSystem.CopyDirectoryRecursive(fullPath, backupPath), $"Backing up {folder} to {backupPath}", out ioError) ||
+                                !this.TryIO(tracer, () => this.fileSystem.DeleteDirectory(fullPath), $"Deleting {fullPath}", out ioError))
+                            {
+                                this.WriteMessage(tracer, $"Backup failed for {folder} and will not be dehydrated. {ioError}");

```

Comment:
It looks like `ioError` will not contain the description of what IO action we were attempting.  Should `TryIO` be updated to include the description in `ioError`?  Or alternatively, should the copy and delete have their own code blocks so that the message we write here can say specifically which step failed?